packages/dds/merge-tree/src/test/client.applyMsg.spec.ts

```diff
@@ -265,147 +265,218 @@ describe("client.applyMsg", () => {
 
     it("conflicting insert over local delete", () => {
-        const clientA = new TestClient();
-        clientA.startOrUpdateCollaboration("A");
-        const clientB = new TestClient();
-        clientB.startOrUpdateCollaboration("B");
-        const clientC = new TestClient();
-        clientC.startOrUpdateCollaboration("C");
-
-        const clients = [clientA, clientB, clientC];
+        const clients = createClientsAtInitialState({ initialState: "" }, "A", "B", "C");
 
         let seq = 0;
         const messages = [
-            clientC.makeOpMessage(clientC.insertTextLocal(0, "CCC"), ++seq),
-            clientC.makeOpMessage(clientC.removeRangeLocal(0, 1), ++seq),
+            clients.C.makeOpMessage(clients.C.insertTextLocal(0, "CCC"), ++seq),
+            clients.C.makeOpMessage(clients.C.removeRangeLocal(0, 1), ++seq),
 
         ];
         while (messages.length > 0) {
             const msg = messages.shift()!;
-            clients.forEach((c) => {
+            clients.all.forEach((c) => {
                 c.applyMsg(msg);
             });
         }
-        const logger = new TestClientLogger(clients);
+        const logger = new TestClientLogger(clients.all);
         logger.validate();
 
         messages.push(
-            clientC.makeOpMessage(clientC.removeRangeLocal(0, 1), ++seq),
-            clientC.makeOpMessage(clientC.insertTextLocal(0, "CC"), ++seq),
-            clientB.makeOpMessage(clientB.insertTextLocal(1, "BBB"), ++seq),
+            clients.C.makeOpMessage(clients.C.removeRangeLocal(0, 1), ++seq),
+            clients.C.makeOpMessage(clients.C.insertTextLocal(0, "CC"), ++seq),
+            clients.B.makeOpMessage(clients.B.insertTextLocal(1, "BBB"), ++seq),
         );
         while (messages.length > 0) {
             const msg = messages.shift()!;
-            clients.forEach((c) => c.applyMsg(msg));
+            clients.all.forEach((c) => c.applyMsg(msg));
+        }
+        logger.validate();
+    });
+
+    it("Local insert after acked local delete", () => {
+        const clients = createClientsAtInitialState(
+            { initialState: "ZZ", options: { mergeTreeUseNewLengthCalculations: true } },
+            "A", "B", "C");
+
+        const logger = new TestClientLogger(clients.all);
+
+        let seq = 0;
+
+        const op1 = clients.C.makeOpMessage(clients.C.removeRangeLocal(0, 1), ++seq);
+        clients.C.applyMsg(op1);
+
+        const op2 = clients.B.makeOpMessage(clients.B.removeRangeLocal(1, 2), ++seq);
+
+        const op3 = clients.C.makeOpMessage(clients.C.insertTextLocal(0, "C"), ++seq);
+
+        const op4 = clients.B.makeOpMessage(clients.B.insertTextLocal(1, "B"), ++seq);
+
+        clients.A.applyMsg(op1);
+        clients.B.applyMsg(op1);
+
+        const messages = [op2, op3, op4];
+        while (messages.length > 0) {
+            const msg = messages.shift()!;
+            clients.all.forEach((c) => c.applyMsg(msg));
+        }
+
+        logger.validate();
+    });
```

Comment:
Am I understanding correctly, that in these tests we don't check against a hardcoded expected final state/string? `logger.validate()` only seems to check that all clients have the same end state as the first one... but do we have any guarantees that that one is right in the first place?

Reply:
Yeah, that's a correct understanding.  we have other test like the replay tests that capture the state and ensure it doesn't diverge over time.

it wouldn't be a bad idea to capture the end state in these cases as we can know it and validate against it. Locally i actually have a change to the logger that takes in a base text. I'll file a bug to get these tests updated once that capability is added.